File: python/jump_game2.py
class Solution(object):
    # O(n) time | O(1) space
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 1:
            return 0
        
        jumps = 1
        maxReach = 0
        reach = nums[0]

        for i in range(1, len(nums) - 1):
            maxReach = max(maxReach, i + nums[i])
            reach -= 1

            if reach == 0:
                jumps += 1
                reach = maxReach - i

        return jumps
    

solution = Solution()
nums = [2,3,1,1,4]

print(solution.jump(nums))

File: src/main/java/january_2026/JumpGame2.java
package january_2026;

public class JumpGame2 {

    public static void main(String[] args) {
        int[] nums = {3,2,1,0,4};
        JumpGame2 jumpGame2 = new JumpGame2();
        boolean result = jumpGame2.canJump(nums);
        System.out.println(result);
    }

    // O(n) time | O(1) space
    public boolean canJump(int[] nums) {
        int destination = nums.length - 1;

        for (int i = nums.length - 1; i >= 0; i--) {
            if (i + nums[i] >= destination) {
                destination = i;
            }
        }

        return destination == 0;
    }

    public boolean canJump2(int[] nums) {
        if (nums.length <= 1) {
            return true;
        }
        if (nums[0] == 0) {
            return false;
        }
        int maxReach = nums[0];

        for (int i = 1; i < nums.length; i++) {
            maxReach = Math.max(maxReach, i + nums[i]);
            if (maxReach >= nums.length - 1) {
                return true;
            }
            if (maxReach <= i) {
                return false;
            }
        }
        return false;
    }

}
